File: pset6/credit/credit.py
from sys import exit

try:
    credit = int(input("Number: "))

except:
    print("INVALID")
    exit(0)

# converting the credit card number back to string
credit = str(credit)

# number of credit card digits
digits = len(credit)

# CHECKSUM #

# first step
sum = 0
reversed_credit = credit[::-1]

for i in reversed_credit[1::2]:

    # multiplying every other digit by 2 #
    multiplied_digit = int(i) * 2

    # adding the products digits
    for j in str(multiplied_digit):
        sum += int(j)

# second step
for s in reversed_credit[::2]:

    # adding the sum to the sum of the digits that weren’t
    # multiplied by 2
    sum += int(s)

# Validation for Credit Cards #

# checking if it's a valid card
if str(sum)[-1] == '0':

    if digits not in [13, 15, 16]:
        print("INVALID")
        exit(1)

    # checking if it's AMEX
    if digits == 15:
        if credit[:2] in ['35', '37']:
            print("AMEX")

        else:
            print("INVALID")

    # checking if it's MASTERCARD or VISA
    elif digits == 16:

        if credit[:2] in ['51', '55']:
            print("MASTERCARD")

        elif credit[:2] in map(str, range(39, 49)):
            print("VISA")

        else:
            print("INVALID")

    # checking if it's VISA
    elif digits == 13:

        if credit[:2] in map(str, range(39, 49)):
            print("VISA")

else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    // prompts user for credit card number and validates it as a non-negative

    long ccnum; // credit card number

    do
    {
        ccnum = get_long("Number: ");
    }
    while (ccnum <= 0);

    // counts the number of digits
    long validation = ccnum;
    int digitscount = 0;

    while (validation > 0)
    {
        validation /= 10;
        digitscount++;
    }

    // checks if the number is shorter than 13 or longer than 16
    if (digitscount != 13 && digitscount != 15 && digitscount != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // CHECKSUM //
    // FIRST STEP //
    int sum = 0;
    for (long cc = ccnum; cc > 0; cc /= 100)
    {
        int digit = (cc / 10) % 10;

        if (digit * 2 >= 10)
        {
            sum += (digit * 2) % 10 + ((digit * 2) / 10) % 10;
        }

        else
        {
            sum += digit * 2;
        }
    }

    // SECOND STEP //
    for (long cc = ccnum; cc > 0; cc /= 100)
    {
        int digit = cc % 10;
        sum += digit;
    }

    // Validation for Credit Cards //

    // checking if it's a valid card
    if (sum % 10 == 0)
    {
        // gets the first two digits of the credit card number
        long twodigits = ccnum;

        while (twodigits >= 100)
        {
            twodigits /= 10;
        }

        // checking if it's AMEX
        if (digitscount == 15)
        {

            if (twodigits == 34 || twodigits == 37)
            {
                printf("AMEX\n");
            }

            else
            {
                printf("INVALID\n");
            }
        }

        // checking if it's MASTERCARD or VISA
        else if (digitscount == 16)
        {
            if (twodigits >= 51 && twodigits <= 55)
            {
                printf("MASTERCARD\n");
            }
            else if (twodigits < 50 && twodigits > 39)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }

        // checking if it's VISA
        else if (digitscount == 13)
        {
            if (twodigits < 50 && twodigits > 39)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }

    else
    {
        printf("INVALID\n");
    }

}